uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -116,476 +76,627 @@
-/// wakers and Pin<> signals that it must not move, since this would break any self-references in
-/// the future.
-pub struct RustFuture<F, T, UT>
+/// # Safety
+///
+/// The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+pub unsafe fn rust_future_cancel(handle: RustFutureHandle) {
+    let future = &*(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    future.clone().ffi_cancel()
+}
+
+/// Complete a Rust future
+///
+/// Note: the actually extern "C" scaffolding functions can't be generic, so we generate one for
+/// each supported FFI type.
+///
+/// # Safety
+///
+/// - The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+/// - The `T` param must correctly correspond to the [rust_future_new] call.  It must
+///   be `<Output as FfiConverter<UT>>::ReturnType`
+pub unsafe fn rust_future_complete<T: FfiDefault>(
+    handle: RustFutureHandle,
+    out_status: &mut RustCallStatus,
+) -> T {
+    let future = &*(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    let mut return_value = T::ffi_default();
+    let out_return = mem::transmute::<&mut T, &mut ()>(&mut return_value);
+    future.ffi_complete(out_return, out_status);
+    return_value
+}
+
+/// Free a Rust future, dropping the strong reference and releasing all references held by the
+/// future.
+///
+/// # Safety
+///
+/// The [RustFutureHandle] must not previously have been passed to [rust_future_free]
+pub unsafe fn rust_future_free(handle: RustFutureHandle) {
+    let future = Box::from_raw(handle.0 as *mut Arc<dyn RustFutureFFI>);
+    future.ffi_free()
+}
+
+/// Thread-safe storage for [RustFutureContinuationCallback] data
+///
+/// The basic guarantee is that all data pointers passed in are passed out exactly once to the
+/// foreign continuation callback. This enables us to uphold the [rust_future_poll] guarantee.
+///
+/// [ContinuationDataCell] also tracks cancellation, which is closely tied to continuation data.
+enum ContinuationDataCell {
+    Empty,
+    Cancelled,
+    Set(*const ()),
+}
+
+impl ContinuationDataCell {
+    fn new() -> Self {
+        Self::Empty
+    }
+
+    /// Store new continuation data
+    fn store(&mut self, data: *const ()) {
+        // If we're cancelled, then call the continuation immediately rather than storing it
+        if let Self::Cancelled = self {
+            call_continuation(data, RustFuturePoll::Ready);
+            return;
+        }
+
+        match mem::replace(self, Self::Set(data)) {
+            Self::Empty => (),
+            Self::Cancelled => unreachable!(),
+            Self::Set(old_data) => {
+                log::error!(
+                    "store: observed Self::Set state, is poll() being called from multiple threads at once?"
+                );
+                call_continuation(old_data, RustFuturePoll::Ready);
+            }
+        }
+    }
+
+    fn send(&mut self) {
+        if matches!(self, Self::Cancelled) {
```

Comment:
For consistency, should also use `if let`, or `fn store` above should also use `matches!`.